plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,304 @@
+        .applyTo(projectSelector);
+    GridLayoutFactory.fillDefaults().spacing(0, 0).generateLayout(projectSelectorComposite);
+
+    setControl(composite);
+
+    gcpIcon = SharedImages.GCP_IMAGE_DESCRIPTOR.createImage();
+    composite.addDisposeListener(new ImageDisposer(gcpIcon));
+  }
+
+  private void updateProjectSelector() {
+    final Credential credential = accountSelector.getSelectedCredential();
+    if (credential == null) {
+      projectSelector.setProjects(new ArrayList<GcpProject>());
+      return;
+    }
+
+    BusyIndicator.showWhile(Display.getCurrent(), new Runnable() {
+      @Override
+      public void run() {
+        try {
+          List<GcpProject> gcpProjects = projectRepository.getProjects(credential);
+          projectSelector.setProjects(gcpProjects);
+        } catch (ProjectRepositoryException e) {
+          logger.log(Level.WARNING,
+              "Could not retrieve GCP project information from server.", e); //$NON-NLS-1$
+        }
+      }
+    });
+  }
+
+  @Override
+  public void setDefaults(ILaunchConfigurationWorkingCopy configuration) {
+    // No particular default values to set in a newly created configuration.
+  }
+
+  @Override
+  public void activated(ILaunchConfigurationWorkingCopy workingCopy) {
+    activated = true;
+    super.activated(workingCopy);
+  }
+
+  @Override
+  public void deactivated(ILaunchConfigurationWorkingCopy workingCopy) {
+    super.deactivated(workingCopy);
+    activated = false;
+
+    if (environmentTab != null) {
+      // Unfortunately, "EnvironmentTab" overrides "activated()" not to call "initializeFrom()".
+      // (Calling "initializeFrom()" when "activated()" is the default implementation of the base
+      // class retained for backward compatibility.) We needed to call it on behalf of
+      // "EnvironmentTab" to re-initialize its UI with the changes made here.
+      environmentTab.initializeFrom(workingCopy);
+    }
+  }
+
+  @Override
+  public void initializeFrom(ILaunchConfiguration configuration) {
+    accountEmailModel = getAttribute(configuration, ATTRIBUTE_ACCOUNT_EMAIL, ""); //$NON-NLS-1$
+
+    Map<String, String> environmentMap = getEnvironmentMap(configuration);
+    gcpProjectIdModel = Strings.nullToEmpty(environmentMap.get(PROJECT_ID_ENVIRONMENT_VARIABLE));
+    String serviceKey = Strings.nullToEmpty(environmentMap.get(SERVICE_KEY_ENVIRONMENT_VARIABLE));
+
+    initializingUiValues = true;
+    // Selecting an account loads projects into the project selector synchronously (via a listener).
+    accountSelector.selectAccount(accountEmailModel);
+    projectSelector.selectProjectId(gcpProjectIdModel);
+    serviceKeyInput.setText(serviceKey);
+    initializingUiValues = false;
+  }
+
+  @Override
+  public void performApply(ILaunchConfigurationWorkingCopy configuration) {
+    if (!activated) {  // to avoid https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2568#discussion_r150128582
```

Comment:
It's probably best to include a potted summary here, in case these projects are moved elsewhere, or Github disappears.

_must avoid updating the environment map unless we're active as performApply() is when the user makes changes in other tabs, like the EnvironmentTab, and thus we could clobber changes made in the EnvironmentTab_